microsoft/testsuites/kselftest/kselftest.py

```diff
@@ -0,0 +1,285 @@
+                cwd=kernel_path,
+                sudo=True,
+            )
+
+            # build kselftests
+            self.node.tools[Make].run(
+                "KBUILD_OUTPUT=build -C tools/testing/selftests gen_tar",
+                cwd=kernel_path,
+                sudo=True,
+                expected_exit_code=0,
+                expected_exit_code_failure_message="could not generate kselftest tar.",
+            ).assert_exit_code()
+
+        tool_path = self.get_tool_path(use_global=True)
+        self.node.tools[Tar].extract(
+            str(self._remote_tar_path), str(tool_path), sudo=True
+        )
+        self._log.debug(f"Extracted tar from path {self._remote_tar_path}!")
+
+        return self._check_exists()
+
+    def run_all(
+        self, test_result: TestResult, environment: Environment, log_path: str
+    ) -> List[KselftestResult]:
+        self.run(
+            # " > kselftest-results.txt",
+            " 2>&1 | tee kselftest-results.txt",
+            force_run=True,
+            sudo=True,
+            shell=True,
+            timeout=5000,
+        )
+
+        # get username
+        username = self.node.tools[Whoami].get_username()
+
+        # Allow read permissions for "others" to remote copy the file
+        # kselftest-results.txt
+        self.node.execute(
+            f"chmod 644 /home/{username}/kselftest-results.txt", sudo=True
+        )
+
+        # copy kselftest-results.txt from remote to local node for processing results
+        remote_copy = self.node.tools[RemoteCopy]
+        remote_copy.copy_to_local(
+            PurePosixPath(f"/home/{username}/kselftest-results.txt"), PurePath(log_path)
+        )
+
+        local_kselftest_results_path = PurePath(log_path) / "kselftest-results.txt"
+
+        # parse results from local_kselftest_results_path file
+        # read the file
+        with open(local_kselftest_results_path, encoding="utf8") as f:
+            result_output = f.read()
+            results = self._parse_results(result_output)
+
+        if not results:
+            raise LisaException("tests did not run, kselftest-results.txt is empty")
+
+        # assert that all tests passed
+        failed_tests = []
+        for result in results:
+            if result.status == TestStatus.FAILED:
+                failed_tests.append(result.name)
+
+            # create test result message
+            info: Dict[str, Any] = {}
+            info["information"] = {}
+            info["information"]["exit_value"] = result.exit_value
+            subtest_message = create_test_result_message(
+                SubTestMessage,
+                test_result,
+                environment,
+                result.name,
+                result.status,
+                other_fields=info,
+            )
+
+            # notify subtest result
+            notifier.notify(subtest_message)
+
+        # assert that none of the tests failed
+        assert_that(failed_tests).is_empty()
```

Comment:
Add `described_as`, so if it's failed here, the message can help find the right code line. I said not to print the array again, but the message is still helpful.